plugins/woocommerce/client/blocks/assets/js/hocs/with-query-loop-product-context-validation.tsx

```diff
@@ -0,0 +1,66 @@
+/**
+ * External dependencies
+ */
+import { __, sprintf } from '@wordpress/i18n';
+import {
+	useBlockProps,
+	Warning,
+	store as blockEditorStore,
+} from '@wordpress/block-editor';
+import { useSelect } from '@wordpress/data';
+import { BlockEditProps } from '@wordpress/blocks';
+
+type EditComponentProps = BlockEditProps< Record< string, never > > & {
+	context: {
+		postType: string;
+	};
+};
+
+/**
+ * Higher-Order Component that checks if a block is inside an invalid Query Loop context.
+ * It renders a warning if the block is inside a Query Loop context and the post type is not a product.
+ *
+ * @param          WrappedComponent The component to wrap
+ * @param {string} blockName        The name of the block to display in the warning message
+ * @return {Function} A wrapped component that includes query loop context validation
+ */
+export const withQueryLoopProductContextValidation = (
+	WrappedComponent: React.ComponentType< EditComponentProps >,
+	blockName: string
+) => {
+	return function WithQueryLoopValidation( props: EditComponentProps ) {
+		const { postType } = props.context;
+		const isInvalidQueryLoopContext = useSelect(
+			( select ) => {
+				const queryLoopAncestors = select(
+					blockEditorStore
+				).getBlockParentsByBlockName(
+					props.clientId,
+					'core/post-template'
+				);
+				return queryLoopAncestors.length > 0 && postType !== 'product';
+			},
+			[ props.clientId, postType ]
+		);
```

Comment:
The old implementation used `useMemo`, just checking if this isn't needed anymore.

```
const { getBlockParentsByBlockName } = useSelect( blockEditorStore, [] );
	const blockParents = useMemo( () => {
		return getBlockParentsByBlockName( clientId, 'core/post-template' );
	}, [ getBlockParentsByBlockName, clientId ] );
```

Reply:
Correct, the old implementation also used `useSelect` in a different way where it just returned the selector name: `const { getBlockParentsByBlockName } = useSelect( blockEditorStore, [] );`
Instead of the recommended way of:
```
const value = useSelect( function( select ) {
  const selectors = select( store );
  const valueFromSelectors = selectors.selector(...);
  return valueFromSelectors;
}, [] );
```
The returned values within the `useSelect` function are memoized, so it uses `useMemo` internally.